Userland/Libraries/LibC/pthread.cpp

```diff
@@ -151,6 +158,23 @@ void pthread_exit(void* value_ptr)
     pthread_exit_without_cleanup_handlers(value_ptr);
 }
 
+void __pthread_maybe_cancel()
+{
+    pthread_t thread = pthread_self();
+
+    // Check if we have cancellations enabled.
+    if (s_thread_cancel_state != PTHREAD_CANCEL_ENABLE)
+        return;
+
+    // Check if a cancellation request is pending.
+    if (!cancellation_requests.contains(thread))
```

Comment:
How is it safe to check a global data structure without any locking or synchronization?

Reply:
Considering that this is used as a set-once type of flag, I don't think races between the cancelling thread and the canceled thread will give much opportunity for issues. It's not necessarily safe, but that doesn't really make it unsafe either. If a request is "missed" it just would end up at some other cancellation point in the future.

Now, if the HashTable itself isn't atomic enough for doing that without locks (is it?), that would be a more important issue. However, if the only solution for that is to introduce locks, I think we should look for another solution entirely, as that seems _very_ slow.

I originally had a TODO there with "Put this in some kind of thread-local storage that everyone can access", but I don't think there is anything like that as of now, right?

Reply:
> Now, if the HashTable itself isn't atomic enough for doing that without locks (is it?), that would be a more important issue. 

This is my main concern, none of the AK data structures are thread safe AFAIK. So a `.contains()` would most certainly race with another thread doing the `.try_set(..)`.

> However, if the only solution for that is to introduce locks, I think we should look for another solution entirely, as that seems very slow. 

Would it be slow in the normal case? You would never get to the hash table check unless a cancel was active. If you want to optimize the check once a cancel is active, maybe you could create a large atomic bitmap of thread ids to avoid needing to do any locking?

Reply:
To answer my own question, I missed that cancel is enabled by default. Oops

Reply:
I replaced the implementation for inter-thread communication with a signal-based one now, that should be atomic enough and well-tested. :^)